src/java.base/share/classes/jdk/internal/jimage/decompressor/StringSharingDecompressor.java

```diff
@@ -95,132 +89,106 @@ public static int[] getSizes() {
 
     @SuppressWarnings("fallthrough")
     public static byte[] normalize(StringsProvider provider, byte[] transformed,
-            int offset) throws IOException {
-        DataInputStream stream = new DataInputStream(new ByteArrayInputStream(transformed,
-                offset, transformed.length - offset));
-        ByteArrayOutputStream outStream = new ByteArrayOutputStream(transformed.length);
-        DataOutputStream out = new DataOutputStream(outStream);
-        byte[] header = new byte[8]; //maginc/4, minor/2, major/2
-        stream.readFully(header);
-        out.write(header);
-        int count = stream.readUnsignedShort();
-        out.writeShort(count);
+                                   int offset, long originalSize) throws IOException {
+        if (originalSize > Integer.MAX_VALUE) {
+            throw new OutOfMemoryError("Required array size too large");
```

Comment:
Is this the correct exception type?

Reply:
> Is this the correct exception type?

I'm not sure, but I think it shouldn't matter, since the exception is never really thrown here for a valid jimage file.

Reply:
It's common to throw `OutOfMemoryError` when a requested array size would be too large, either due overflow or because some other limitation (see `StringUTF16::newBytesFor` or `String::replace` for some examples).